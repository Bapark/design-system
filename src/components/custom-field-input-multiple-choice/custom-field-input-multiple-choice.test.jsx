import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  screen,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CustomFieldInputMultipleChoice from './custom-field-input-multiple-choice.jsx';

describe('<CustomFieldInputMultipleChoice>', () => {
  const requiredProps = {
    choices: [{
      id: '1',
      label: 'Choice 1',
    }, {
      id: '2',
      label: 'Choice 2',
    }],
    id: 'test-id',
    label: 'test label',
  };

  it('has defaults', () => {
    expect(renderer.create(<CustomFieldInputMultipleChoice {...requiredProps} />).toJSON()).toMatchSnapshot();
  });

  describe('deselecting a choice', () => {
    it('removes the choice value when pressing the remove button', () => {
      render((<CustomFieldInputMultipleChoice
        {...requiredProps}
        value={requiredProps.choices}
      />));

      expect(screen.getByText('Choice 1')).toBeInTheDocument();
      expect(screen.getByText('Choice 2')).toBeInTheDocument();

      userEvent.click(screen.getAllByRole('button')[0]);
      expect(screen.queryByText('Choice 1')).not.toBeInTheDocument();
      expect(screen.getByText('Choice 2')).toBeInTheDocument();
    });
  });

  describe('choices API', () => {
    it('can be set', () => {
      render((<CustomFieldInputMultipleChoice
        {...requiredProps}
        choices={[{
          id: '42',
          label: 'Answer to everything',
        }]}
      />));

      expect(screen.queryByRole('option', 'Answer to everything')).not.toBeInTheDocument();
      userEvent.click(screen.getByLabelText('test label'));
      expect(screen.getByRole('option', 'Answer to everything')).toBeInTheDocument();
    });

    it('does not show a selected choice', () => {
      const choices =[{
        id: '42',
        label: 'Answer to everything',
      }]
      render((<CustomFieldInputMultipleChoice
        {...requiredProps}
        choices={choices}
        value={choices}
      />));

      userEvent.click(screen.getByLabelText('test label'));
      expect(screen.queryByRole('option', 'Answer to everything')).not.toBeInTheDocument();
    });
  });

  describe('id API', () => {
    it('generates unique tag ids', () => {
      render((<CustomFieldInputMultipleChoice
        {...requiredProps}
        id="123"
        value={requiredProps.choices}
      />));

      expect(screen.getByText('Choice 1').parentElement).toHaveAttribute('id', '123-1');
      expect(screen.getByText('Choice 2').parentElement).toHaveAttribute('id', '123-2');
    });
  });

  describe('label API', () => {
    it('can be set', () => {
      render((<CustomFieldInputMultipleChoice {...requiredProps} label="Unique label" />));

      expect(screen.getByLabelText('Unique label')).toBeInTheDocument();
    });
  });

  describe('readOnly API', () => {
    it('does not have the clear button on its tags', () => {
      render((<CustomFieldInputMultipleChoice
        {...requiredProps}
        readOnly={true}
        value={[requiredProps.choices[0]]}
      />));

      expect(screen.queryByRole('button')).toBeNull();
    });

    it('has the clear button on its tags', () => {
      render((<CustomFieldInputMultipleChoice
        {...requiredProps}
        readOnly={false}
        value={[requiredProps.choices[0]]}
      />));

      expect(screen.getByRole('button')).toBeInTheDocument();
    });
  });

  describe('value API', () => {
    it('generates tags', () => {
      render((<CustomFieldInputMultipleChoice
        {...requiredProps}
        value={requiredProps.choices}
      />));

      expect(screen.getByText('Choice 1')).toBeInTheDocument();
      expect(screen.getByText('Choice 2')).toBeInTheDocument();
    });
  });
});
