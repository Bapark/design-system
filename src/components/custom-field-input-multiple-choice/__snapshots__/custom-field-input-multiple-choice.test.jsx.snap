// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomFieldInputMultipleChoice> has defaults 1`] = `
<div>
  <div
    onKeyDown={[Function]}
    role="presentation"
  >
    <label
      className="label"
      htmlFor="test-id-autocomple"
      id="test-id-label"
    >
      test label
    </label>
    <div
      className="control-container"
    >
      <div
        className="read-write-container"
        onClick={[Function]}
        role="presentation"
      >
        <div
          aria-labelledby="test-id-label"
          className="tag-list"
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="grid"
        >
          <input
            aria-haspopup="listbox"
            aria-labelledby="test-id-label"
            className="autocomplete-input"
            id="test-id-autocomple"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          className="icons-container"
        >
          <svg
            className="icon size-medium stroke-none color-transparent"
            onClick={[Function]}
            role="img"
            tabIndex={-1}
          >
            <use
              onKeyDown={[Function]}
              xlinkHref="#icon-caret-down.svg"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
